pkg/lib/csv/replace_finder.go

```diff
@@ -0,0 +1,65 @@
+package csv
+
+import (
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NewReplaceFinder returns an instance of ReplaceFinder
+func NewReplaceFinder(logger *logrus.Logger, client versioned.Interface) ReplaceFinder {
+	return &replace{
+		logger: logger,
+		client: client,
+	}
+}
+
+// ReplaceFinder is an interface that finds the next or previous
+// ClusterServiceVersion object in the upgrade path for a given CSV.
+type ReplaceFinder interface {
+	IsBeingReplaced(in *v1alpha1.ClusterServiceVersion, csvsInNamespace map[string]*v1alpha1.ClusterServiceVersion) (replacedBy *v1alpha1.ClusterServiceVersion)
+	IsReplacing(in *v1alpha1.ClusterServiceVersion) *v1alpha1.ClusterServiceVersion
+}
+
+type replace struct {
+	logger *logrus.Logger
+	client versioned.Interface
+}
+
+// IsBeingReplaced returns the corresponding ClusterServiceVersion object that
+// is replacing the given CSV specified.
+//
+// If the corresponding ClusterServiceVersion is not found nil is returned.
+func (r *replace) IsBeingReplaced(in *v1alpha1.ClusterServiceVersion, csvsInNamespace map[string]*v1alpha1.ClusterServiceVersion) (replacedBy *v1alpha1.ClusterServiceVersion) {
+	for _, csv := range csvsInNamespace {
+		r.logger.Infof("checking %s", csv.GetName())
+		if csv.Spec.Replaces == in.GetName() {
+			r.logger.Infof("%s replaced by %s", in.GetName(), csv.GetName())
+			replacedBy = csv.DeepCopy()
```

Comment:
It would be nice to review where we're calling this and see if the copy is really necessary. This can be quite a lot of objects in a larger cluster.

Another thought: we might want to filter out Copied CSVs here - but again that will require a review of where we call this.

Reply:
I was thinking whether deep copy was necessary for this but didn't spend time since I thought it would be outside the scope of this PR.
But I am happy to do the search and see whether we can filter out copied CSVs and not deep copy.